### PR TITLE
Unit test attributes

### PR DESCRIPTION
**Pull request by Quiescent**
# Test and Fix Attributes for Characters
The new attributes feature for characters touched some ancient code.
In doing so, it unearthed curious bugs compounded by the recent addition of per-book attributes.
This PR adds 92 test cases to `pltr` and fixes a few bugs found along the way.

## `characters.test.js`
 ```
   editCharacterTemplateAttribute
    given a store with no characters
      ✓ should leave the state unchanged (12 ms)
    given a store with a character
      but no templates applied to that character
        ✓ should leave the state unchanged (3 ms)
      a template applied to that character
        and the id of an attribute in that template
          ✓ should change the value of that attribute (4 ms)
        and an id for an attribute *not* in that template
          ✓ should leave the state unchanged (3 ms)
        the id of an attribute that *is* in the template
          when the attribute is not yet modified
            ✓ should add that attribute (2 ms)
          when the attribute is already there
            ✓ should modify that attribute (2 ms)
          when the attribute is in another book
            if tabs should not be shown
              ✓ should modify the all version of the attribute (3 ms)
            if tabs should be shown
              ✓ should add another template value for the current book (3 ms)
  addTag
    given a store with no characters and no tags
      ✓ should add a base attribute for tags to the attributes collection (1 ms)
    given a store with a character
      and no tags
        ✓ should a base attribute for tags to the attributes collection (1 ms)
      and only one book
        when book one tab is selected
          ✓ should apply the tag to the "all" book (2 ms)
        when the "all" book tab is selected
          ✓ should apply the tag to the "all" book (1 ms)
      and two books
        when no book is associated with the character
          ✓ should add the tag to the "all" book (2 ms)
        when a book is associated with the character
          and the  "all" tab is selected
            ✓ should add the tag to the "all" book (2 ms)
          and the first book is selected
            ✓ should add the tag to the first book (3 ms)
  addBook
    given a state with no characters
      ✓ leave the state un changed (1 ms)
    given a state with a character
      and a different character id
        ✓ should leave the state unchanged (1 ms)
      and that characters id
        ✓ should add the book to that character (1 ms)
        and a book id that does not exist
          ✓ should leave the state unchanged (1 ms)
  removeTag
    given a state with no characters
      ✓ should leave the state unchanged (1 ms)
    given a state with a character
      and no tags
        ✓ should leave the state unchanged (1 ms)
      and a tag
        but the tag is not associated with the character
          ✓ should leave the state unchanged (1 ms)
        and the tag is associated with the character
          ✓ should remove the tag from the character (2 ms)
          when the tag is associated in both book 1 and "all"
            and book 1 is selected
              ✓ should remove the tag from book 1 (4 ms)
            and the "all" book is selected
              ✓ should remove the tag from the "all" book (4 ms)
  removeBook
    given a state with no characters
      ✓ leave the state un changed (4 ms)
    given a state with a character
      and no books associated with that character
        ✓ should leave the state unchanged (1 ms)
      and a book associated with that character
        and that book id
          ✓ should remove it from the character (1 ms)
        and a different book id
          ✓ should leave the state un changed (1 ms)
      and that characters id
        ✓ should add the book to that character (1 ms)
        and a book id that does not exist
          ✓ should leave the state unchanged
  createCharacterAttribute
    given a state with no characters
      ✓ should not change the character state (1 ms)
    given a state with a character
      ✓ should and and default the value of the new attribute to the character (1 ms)
  editCharacterAttributeValue
    given a state with no characters
      ✓ should leave the state unchanged (1 ms)
    given a state with a character
      and no attributes
        ✓ should leave the state unchanged (1 ms)
      and a character attribute
        and a different character id
          ✓ should leave the character state unchanged (1 ms)
      and that character id
        and an attribute
          and a different attribute id
            ✓ should leave the state unchanged
          and that attribute id
            ✓ should edit that attribute value (1 ms)
  editShortDescription
    given a store with no characters
      ✓ should add an attribute but leave the character state unchanged
    given a store with a character
      and a different character id
        ✓ should leave the character state unchanged and create an attribute for the description (1 ms)
      and that character id
        ✓ should edit the character short description for the all book (1 ms)
        and more than one book
          when no book is associated with a character
            ✓ should edit the short description for the "all" book (1 ms)
          when a book is associated with a character
            and the all book is selected
              ✓ should edit the short description for the all book (1 ms)
            and the first book is selected
              ✓ should edit the short description for the first book (2 ms)
  editDescription
    given a store with no characters
      ✓ should add an attribute but leave the character state unchanged (1 ms)
    given a store with a character
      and a different character id
        ✓ should leave the character state unchanged and create an attribute for the description (1 ms)
      and that character id
        ✓ should edit the character short description for the all book (1 ms)
        and more than one book
          when no book is associated with a character
            ✓ should edit the short description for the "all" book (1 ms)
          when a book is associated with a character
            and the all book is selected
              ✓ should edit the short description for the all book (1 ms)
            and the first book is selected
              ✓ should edit the short description for the first book (2 ms)
  editCategory
    given a store with no characters
      ✓ should add an attribute but leave the character state unchanged (1 ms)
    given a store with a character
      and a different character id
        ✓ should leave the character state unchanged and create an attribute for the description (1 ms)
      and that character id
        ✓ should edit the character short description for the all book (1 ms)
        and more than one book
          when no book is associated with a character
            ✓ should edit the short description for the "all" book (1 ms)
          when a book is associated with a character
            and the all book is selected
              ✓ should edit the short description for the all book (2 ms)
            and the first book is selected
              ✓ should edit the short description for the first book (1 ms)
  deleteBook
    given a state with a character
      that is not associated with any books
        ✓ should leave the character state unchanged (1 ms)
      that is associated with a book
        and that book id
          ✓ should remove the association with the deleted book (2 ms)
        and a different book id
          ✓ should leave the state unchanged (1 ms)
  deleteCharacterCategory
    given a state with no characters
      ✓ should leave the character state unchanged (1 ms)
    given a state with a character
      but that character has no category
        ✓ should leave the state unchanged (1 ms)
      and the character has a category
        and that category id
          ✓ should remove the association between the character and category (1 ms)
          and we are in a different book tab
            ✓ should still remove the association (2 ms)
        and a different category id
          ✓ should leave the character unchanged (1 ms)
  deleteTag
    given a state with no characters
      ✓ should leave the character state unchanged (1 ms)
    given a state with a character
      but that character has no tags
        ✓ should leave the state unchanged (1 ms)
      and the character has a tag
        and that tag id
          ✓ should remove the association between the character and tag (1 ms)
          and we are in a different book tab
            ✓ should still remove the association (2 ms)
        and a different tag id
          ✓ should leave the character unchanged (1 ms)
  deleteCharacterAttribute
    given a state with no characters
      ✓ should leave the character state unchanged (1 ms)
    given a state with a character
      but no custom attributes
        ✓ should leave the character state unchanged (1 ms)
      and a custom attribute
        and that custom attribute id
          ✓ should remove the attribute for that custom attribute entirely (1 ms)
          and a different book is selected
            ✓ should still remove the attribute for that custom attribute entirely (1 ms)
        and a different custom attribute id
          ✓ should leave the character state unchanged (1 ms)

Test Suites: 1 passed, 1 total
Tests:       74 passed, 74 total
```

## `attributes.test.js`

```
  editCharacterAttributeMetadata
    given a store with no characters
      ✓ should leave the state unchanged (13 ms)
    given a store with a character in it
      and no character attributes
        ✓ should leave the state unchanged (3 ms)
      and a character attribute
        ✓ should change the details of that attribute (4 ms)
        ✓ should not change the details of that attribute if the id is different (3 ms)
  deleteCharacterAttribute
    given a store with no characters
      ✓ should leave the state unchanged (1 ms)
    given a store with a character in it
      and no character attributes
        ✓ should leave the state unchanged (2 ms)
      and a character attribute
        ✓ should delete that attribute (2 ms)
        ✓ should not delete that attribute if the id is different (2 ms)
  reorderCharacterAttribute
    given a store with no characters
      ✓ should leave the state unchanged (1 ms)
    given a store with a character in it
      and no character attributes
        ✓ should leave the state unchanged (2 ms)
      and a character attribute
        ✓ should leave that attribute unchanged when the target index is 0 (1 ms)
        ✓ should leave that attribute unchanged when the target index is 1 (2 ms)
        ✓ should leave that attribute unchanged when the target index is -1 (1 ms)
        ✓ should not reorder that attribute if the given attribute is different (2 ms)
      and a two character attributes
        ✓ should leave that attribute unchanged when the target index is 0 (2 ms)
        ✓ should move that attribute when the target index is 1 (2 ms)
        ✓ should leave that attribute unchanged when the target index is -1 (1 ms)
        ✓ should not reorder that attribute if the given attribute is different (3 ms)

Test Suites: 1 passed, 1 total
Tests:       18 passed, 18 total
```

### CODE REVIEW

File: lib/pltr/v2/reducers/root.js [hunk 1]
```diff
@@ -119,7 +130,8 @@ const root = (dataRepairers) => (state, action) => {
       )
       return mainReducer(state, newAction)
     }
-    case ATTACH_TAG_TO_CHARACTER: {
+    case ATTACH_TAG_TO_CHARACTER:
```
Comment (jeanawhou) [inline lib/pltr/v2/reducers/root.js:133]: @Quiescent im not sure why `ATTACH_TAG_TO_CHARACTER` and `DELETE_TAG` has the same reducer :sweat_smile:

Comment (Quiescent) [reply lib/pltr/v2/reducers/root.js:133]: That's in the root reducer.  In that reducer, we add some extra detail to various actions before they reach specific reducers.
In the case of these new attributes, the downstream reducer needs to know whether we needed to make a new attribute, what the current book id is and sometimes what all the available attribute ids are.

Comment (jeanawhou) [reply lib/pltr/v2/reducers/root.js:133]: Ah they just modify the base attribute :sweat_smile: 
I tested it and it remove the old tags of the character

**Review by cameronsutter — COMMENTED** [review]
LGTM

**Pull request merged**